esp-hal-common/src/embassy/time_driver_systimer.rs

```diff
@@ -83,17 +91,20 @@ impl EmbassyTimer {
         alarm: embassy_time::driver::AlarmHandle,
         timestamp: u64,
     ) -> bool {
-        critical_section::with(|cs| {
+        critical_section::with(|_cs| {
             let n = alarm.id() as usize;
-            let alarm_state = &self.alarms.borrow(cs)[n];
 
             // The hardware fires the alarm even if timestamp is lower than the current
-            // time.
-            alarm_state.timestamp.set(timestamp);
+            // time. In this case the interrupt handler will pend a wakeup when we exit the
+            // critical section.
             self.arm(n, timestamp);
+        });
 
-            true
-        })
+        // In theory, the above comment is true. However, in practice, we seem to be
```

Comment:
This might be a problem with the systimer driver instead, we may have to clear the `SYSTIMER_TARGETx_WORK_EN` bit whilst loading the alarm value, as for short targets the timer may run past the value by the time it is loaded?